### ISSUE TITLE
Modular LMMS

### ISSUE THREAD

**Issue by BrahRah** [LMMS/lmms#6272]
This would allow users to use the UI layout they want.

Instead of making a single window layout like #1911 OR detachable widows like #3532 the user uses a workspace preset or makes his own. 

An empty workspace could  look like this:
![Empty Workspace](https://user-images.githubusercontent.com/40721994/149316597-89a02382-b438-40be-8e5f-5043f7b28625.png)

Anything would be split into a module that works on it's own:
![Modules](https://user-images.githubusercontent.com/40721994/149316867-1a1714e6-7f83-42be-8ec8-2cd27378bbd6.png)

Every Module can control and receive input. Like the Instrument Plugin opens a new workspace window and the toolbar opens and collapses the Plugins-Browser. The difference however is that the receiving end can be changed. A user could specify that the Plugins-Browser changes the Single-Window-Workspaces shown module or opens a new workspace window inside a infinite workspace that resides on a different monitor.

Basically every UI suggestion could have it's own application workspace tab:
![budislav workspace](https://user-images.githubusercontent.com/40721994/149319940-89534956-f895-4347-a0e5-7657b57cd2be.png)
The budislav single workspace would consists of 4 workspace modules. The top one using a fixed toolbar. The left one containing a browse tab that can be switched out with other workspace tabs. The middle  one a workspace that is further split in two with further tabs for top and bottom. The right one containing a workspace module tab that is dedicated to the controller.

Here are some more possible application workspace tabs:
![InfiniteWorkspaceOnly](https://user-images.githubusercontent.com/40721994/149320927-333f6c23-6bd4-4905-a7ed-664e8a86a487.png)
This is workspace has no modules that are fixed on the screen. Every module is placed inside the infinite canvas and the user moves by panning, shortkeys and module buttons that change the focus on the module the user wants to work on.

![FixedToolbars with infinite workspace](https://user-images.githubusercontent.com/40721994/149321382-53f2f013-26c7-467e-a610-9233b9dfad67.png)
This workspace has some slim fixed to screen modules for for the focus but everything else is placed onto the infinite workspace. 

![CustomModularWorkspace](https://user-images.githubusercontent.com/40721994/149321010-3a43c116-947a-49ae-b0b1-c0ef3e9d76dc.png)
This shows how a user could create a custom workspace.

If it's still not completely clear every workspace itself can be filled with modules to use. That workspace itself becomes a module.
This way a small controller module, the keyboard or midi device can be used to control aspects of the workspace or the modules inside the workspace. Those that are familiar with how adobe animate works basically everything becomes a symbol/container.

**Comment by Monospace-V** [comment]
Asking for https://github.com/LMMS/lmms/pull/3532 detachable windows, to confirm? Combined with UI rearranging and tabbed docking?

**Comment by BrahRah** [comment]
Yes and no the layout would work without detachable windows/modules but it's def. better to have detachable/movable windows/modules. In my suggestion you don't really move them outside the workspace you just rearrange them inside one or many different work-spaces. Having them move outside the work-spaces would compliment my interface suggestion tho.

If anything is anything unclear just mention the part and I'll rewrite or add to it.

**Comment by Monospace-V** [comment]
like the windows of N++, then?

**Comment by BrahRah** [comment]
Questions are always welcome nothing to be sorry about. If things are unclear ask away!
I'd say more like blender. There you have workspace tabs at the top and you can split the screen into different areas. 
Every area has a different functionality.

You can imagine my concept like a desk (workspace), with different tools (windows/modules). 
There are also different desks with mainly the same tools but the tools are placed differently on each desk.
You choose which desk to use and you can even create your own desk with different tool placement.

**Comment by Lazerbeak12345** [comment]
I've seen a lot of GUIs that can do things like this over the years. The best I've seen so far has been Inkscape's new panels. If you drag something somewhere, it goes there. If there's a place it can go, it's highlighted. If you leave the window or place it in an invalid place, it becomes a new window. There are handles for resizing, you can attach multiple floating windows, and more.

I'm of course not saying that we should _clone_ inkscape, but I am saying that their approach is by far the best I've used, and should be considered in the design. They focus on highlighting possible actions before the user takes them, and as a user, it just makes sense.

(important) Differences between Inkscape's modular panels and Blender's modular panels for LMMS to consider:

- Inkscape's panels are always tab collections (which can be merged)
- Inkscape's panels have a drag handle (one as the tab button, the other as the tab bar for the whole tab collection) to change where the panel is. When dragging, available targets are highlighted (and blink/fade/react before impending action)
- Blender has a context menu in the header that lets you split, pop out, duplicate, and more.
  - Note that it has a _duplicate_ feature. In LMMS, this might be a must-have.
- Inkscape does not have workspaces, but Blender's workspaces are hard to navigate (for newbs) since there's no GUI indicator or selector in the default settings (there's a tab bar available, older versions had it on by default).

That all said, what do you think? @BrahRah have you taken a look at Inkscape's panels (a new feature for Inkscape in version 1-ish)? What are your thoughts?
